services/libs/aster-std/src/process/vm/mod.rs

```diff
@@ -0,0 +1,164 @@
+//!
+//! The `Vm` struct contains `Vmar`, which stores all existing memory mappings. The `Vm` also contains
+//! the basic info of process level vm segments, like init stack and heap.
+//!
+//! TODO: The vm layout should be redesigned in the future. Currently, we only use a small part of the
+//! whole virtual memory space.
+//!
+
+mod heap;
+mod stack;
+
+use super::constants::{MAX_ARG_LEN, MAX_ENVP_NUMBER, MAX_ENV_LEN};
+use crate::prelude::*;
+use crate::vm::vmar::Vmar;
+use alloc::ffi::CString;
+use aster_frame::vm::VmIo;
+use aster_rights::Full;
+
+pub use heap::Heap;
+pub use stack::Stack;
+
+/*
+* The user space virtual memory layout looks like below.
+* |-----------------------|-------The highest user vm address
+* |                       |
+* |       Mmap Areas      |
+* |                       |
+* |                       |
+* --------------------------------The init stack base
+* |                       |
+* | User Stack(Init Stack)|
+* |                       |
+* |         ||            |
+* ----------||----------------------The user stack top, grows down
+* |         \/            |
+* |                       |
+* |     Unmapped Areas    |
+* |                       |
+* |         /\            |
+* ----------||---------------------The user heap top, grows up
+* |         ||            |
+* |                       |
+* |        User Heap      |
+* |                       |
+* ----------------------------------The user heap base
+*/
+
+/// The process user space virtual memory
+pub struct Vm {
+    root_vmar: Vmar<Full>,
+    heap: Heap,
+    stack: Stack,
+}
+
+impl Clone for Vm {
+    fn clone(&self) -> Self {
+        Self {
+            root_vmar: self.root_vmar.dup().unwrap(),
+            heap: self.heap.clone(),
+            stack: self.stack.clone(),
+        }
+    }
+}
+
+impl Vm {
+    pub fn alloc() -> Self {
+        let root_vmar = Vmar::<Full>::new_root();
+        let heap = Heap::new();
+        let stack = Stack::new();
+        heap.init(&root_vmar);
+        Vm {
+            root_vmar,
+            heap,
+            stack,
+        }
+    }
+
+    pub fn new(root_vmar: Vmar<Full>, heap: Heap, stack: Stack) -> Self {
+        Self {
+            root_vmar,
+            heap,
+            stack,
+        }
+    }
+
+    pub fn root_vmar(&self) -> &Vmar<Full> {
+        &self.root_vmar
+    }
+
+    pub(super) fn stack(&self) -> &Stack {
+        &self.stack
+    }
+
+    pub fn heap(&self) -> &Heap {
+        &self.heap
+    }
+
+    /// Set the `Vm` to the init status
+    pub fn clear(&self) {
+        self.root_vmar.clear().unwrap();
+        self.heap.set_default(&self.root_vmar);
+        self.stack.clear();
+    }
+
+    /// Read argc from the process init stack
+    pub fn argc(&self) -> Result<u64> {
+        let stack_bottom = self.stack.bottom()?;
+        Ok(self.root_vmar.read_val(stack_bottom)?)
+    }
+
+    /// Read argv from the process init stack
+    pub fn argv(&self) -> Result<Vec<CString>> {
```

Comment:
I suggest moving `argc`, `argv`, and `envp` to `Stack`. And `Stack` should keep a child slice of the root `Vmar`.  With this refactoring, `Stack` would become a better abstraction.

Reply:
`Vmar` cannot create child slice now. So `Stack` contains a field `Vmar` which represents the process vm.